File: backend-java/src/main/java/com/pdfdiff/service/ResultMapper.java
package com.pdfdiff.service;

import com.pdfdiff.domain.LineUnit;
import com.pdfdiff.domain.RawChange;
import com.pdfdiff.dto.ChangeItem;
import com.pdfdiff.dto.CompareResult;
import com.pdfdiff.dto.CompareSummary;
import com.pdfdiff.dto.LineInfo;
import com.pdfdiff.dto.SideInfo;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class ResultMapper {

    public CompareResult buildCompareResult(
            String jobId,
            List<LineUnit> templateLines,
            List<LineUnit> contractLines,
            List<RawChange> rawChanges
    ) {
        List<ChangeItem> changes = new ArrayList<>();
        CompareSummary summary = new CompareSummary();
        int changeIndex = 0;

        for (RawChange raw : rawChanges) {
            if ("equal".equals(raw.type())) {
                summary = new CompareSummary(
                        summary.deletedLines(),
                        summary.insertedLines(),
                        summary.modifiedLines(),
                        summary.equalLines() + 1
                );
                continue;
            }

            changeIndex++;
            ChangeItem item = toChangeItem(String.format("c%04d", changeIndex), raw);
            changes.add(item);
            summary = updateSummary(summary, item);
        }

        return new CompareResult(
                jobId,
                summary,
                changes,
                templateLines.stream().map(this::toLineInfo).toList(),
                contractLines.stream().map(this::toLineInfo).toList()
        );
    }

    private ChangeItem toChangeItem(String changeId, RawChange raw) {
        return new ChangeItem(
                changeId,
                raw.type(),
                sideFromLines(raw.templateLines()),
                sideFromLines(raw.contractLines())
        );
    }

    private SideInfo sideFromLines(List<LineUnit> lines) {
        if (lines == null || lines.isEmpty()) {
            return null;
        }
        LineUnit line = lines.get(0);
        double[] bbox = line.bbox();
        return new SideInfo(
                line.page(),
                line.text(),
                List.of(List.of(bbox[0], bbox[1], bbox[2], bbox[3]))
        );
    }

    private LineInfo toLineInfo(LineUnit line) {
        double[] bbox = line.bbox();
        return new LineInfo(
                line.id(),
                line.page(),
                line.text(),
                List.of(List.of(bbox[0], bbox[1], bbox[2], bbox[3]))
        );
    }

    private CompareSummary updateSummary(CompareSummary summary, ChangeItem item) {
        return switch (item.type()) {
            case "delete" -> new CompareSummary(
                    summary.deletedLines() + 1,
                    summary.insertedLines(),
                    summary.modifiedLines(),
                    summary.equalLines()
            );
            case "insert" -> new CompareSummary(
                    summary.deletedLines(),
                    summary.insertedLines() + 1,
                    summary.modifiedLines(),
                    summary.equalLines()
            );
            case "replace" -> new CompareSummary(
                    summary.deletedLines(),
                    summary.insertedLines(),
                    summary.modifiedLines() + 1,
                    summary.equalLines()
            );
            default -> summary;
        };
    }
}

File: frontend/src/lib/compare/resultMapper.ts
import type { ChangeItem, CompareResult, LineInfo, PdfPageSize } from '../../types/compare'
import type { LineUnit, RawChange } from './types'

/** 将 RawChange 的单侧行 + bbox 转为 ChangeItem 的 template/contract 字段 */
function sideFromLines(
  lines: LineUnit[],
  bboxesOverride: number[][] | null,
): ChangeItem['template'] {
  if (!lines.length || !bboxesOverride?.length) return null
  const line = lines[0]
  return { page: line.page, text: line.text, bboxes: bboxesOverride }
}

/** RawChange → ChangeItem，生成 c0001 格式 id */
function toChangeItem(changeId: string, raw: RawChange): ChangeItem {
  return {
    id: changeId,
    type: raw.type as ChangeItem['type'],
    level: 'char',
    template: sideFromLines(raw.templateLines, raw.templateBboxes),
    contract: sideFromLines(raw.contractLines, raw.contractBboxes),
  }
}

/** LineUnit → API 响应中的 LineInfo */
function toLineInfo(line: LineUnit): LineInfo {
  return {
    id: line.id,
    page: line.page,
    text: line.text,
    bboxes: [line.bbox],
  }
}

/** 将 RawChange 列表转为 UI 使用的 CompareResult，并统计变更摘要 */
export function buildCompareResult(
  jobId: string,
  templateLines: LineUnit[],
  contractLines: LineUnit[],
  rawChanges: RawChange[],
  templatePageSizes?: PdfPageSize[],
  contractPageSizes?: PdfPageSize[],
): CompareResult {
  const changes: ChangeItem[] = []
  let deletedLines = 0
  let insertedLines = 0
  let modifiedLines = 0
  let equalLines = 0
  let changeIndex = 0

  for (const raw of rawChanges) {
    if (raw.type === 'equal') {
      equalLines++
      continue
    }

    changeIndex++
    const item = toChangeItem(`c${String(changeIndex).padStart(4, '0')}`, raw)
    changes.push(item)

    if (item.type === 'delete') deletedLines++
    else if (item.type === 'insert') insertedLines++
    else if (item.type === 'replace') modifiedLines++
  }

  return {
    job_id: jobId,
    status: 'done',
    summary: {
      deleted_lines: deletedLines,
      inserted_lines: insertedLines,
      modified_lines: modifiedLines,
      equal_lines: equalLines,
    },
    changes,
    template_lines: templateLines.map(toLineInfo),
    contract_lines: contractLines.map(toLineInfo),
    template_page_sizes: templatePageSizes,
    contract_page_sizes: contractPageSizes,
  }
}
